Share the implementation of hashing for both vars_prompt and password_hash

mfuchs opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
vars_prompt
filter

##### ANSIBLE VERSION
```
2.2.1
```

##### SUMMARY

Shares the implementation of hashing secrets for both vars_prompt and password_hash.

* vars_prompt with encrypt does not require passlib for the algorithms
  supported by crypt.
* Additional checks ensure that there is always a result.
  This works around issues in the crypt.crypt python function that returns
  None for algorithms it does not know.
  Some modules (like user module) interprets None as no password at all,
  which is misleading.
* The password_hash filter supports all parameters of passlib.
  This allows users to provide a rounds parameter, fixing #15326.
* password_hash is not restricted to the subset provided by crypt.crypt,
  fixing one half of #17266.
* Updated documentation fixes other half of #17266.
* password_hash does not hard-code the salt-length, which fixes bcrypt
  in connection with passlib.
  bcrypt requires a salt with length 22.
* Salts are only generated by ansible when using crypt.crypt.
  Otherwise passlib generates them.
* Avoids deprecated functionality of passlib with newer library versions.

Fixes #15326
Fixes #17266

<!-- Paste verbatim command output below, e.g. before and after your change -->
```
# after
{{ 'secret' | password_hash('bcrypt', rounds=14) }}
```

mattclay commented:
CI failure due to unit test errors. Here's one of the errors:

```
2017-02-09 23:40:23 =================================== FAILURES ===================================
2017-02-09 23:40:23 _________________________ TestEncrypt.test_do_encrypt __________________________
2017-02-09 23:40:23 
2017-02-09 23:40:23 self = <units.utils.test_encrypt.TestEncrypt testMethod=test_do_encrypt>
2017-02-09 23:40:23 
2017-02-09 23:40:23     def test_do_encrypt(self):
2017-02-09 23:40:23         self.assertTrue(encrypt.PASSLIB_AVAILABLE)
2017-02-09 23:40:23     
2017-02-09 23:40:23 >       with self.assertRaises(AnsibleError):
2017-02-09 23:40:23 E       TypeError: failUnlessRaises() takes at least 3 arguments (2 given)
2017-02-09 23:40:23 
2017-02-09 23:40:23 test/units/utils/test_encrypt.py:76: TypeError
```

mattclay commented:
CI failure in unit tests due to:

```
2017-02-10 21:37:00 E           MissingBackendError: bcrypt: no backends available -- recommend you install one (e.g. 'pip install bcrypt')
```

The full error is quite long, so I'll just link to the logs here: https://app.shippable.com/runs/589e2dee031d631000bae9d4/2/console

mfuchs commented:
Thanks mattclay for your input! :)

ready_for_review

abadger commented:
Catch and reraise is a better pattern than passing the exception in.  Python 2 makes this a pain, though.    Ned Batchelder has a good blog post about how to do that right: https://nedbatchelder.com/blog/200711/rethrowing_exceptions_in_python.html

passlib changes number of rounds while crypt has a static value.  I'm not sure of al the ramifications for that.  It would mean that pre and post change, lookup('password_hash'[...]) will return different strings.  If rounds is calculated dynamically at runtime, it could mean that when I use a playbook that has lookup('password_hash'[...]) on my low end laptop, I generate a different string than when someone else runs it on their high end workstation.  If that's true it leads to non-repeatability of playbook runs.

mfuchs commented:
As discussed with abadger I will make sure that password_hash returns the same value for both 2.2 and when this commit is merged.
No matter if passlib is installed or not.
This will also be part of the unit test.

Furthermore I will look into allowing a rounds parameter for crypt, for example crypt.crypt('secret', '$6$rounds=656000$abc').

dharmabumstead reviewed (approved):
Doc updates are OK. Thanks @mfuchs!

mfuchs committed:
Share the implementation of hashing for both vars_prompt and password_hash.

* vars_prompt with encrypt does not require passlib for the algorithms
  supported by crypt.
* Additional checks ensure that there is always a result.
  This works around issues in the crypt.crypt python function that returns
  None for algorithms it does not know.
  Some modules (like user module) interprets None as no password at all,
  which is misleading.
* The password_hash filter supports all parameters of passlib.
  This allows users to provide a rounds parameter, fixing #15326.
* password_hash is not restricted to the subset provided by crypt.crypt,
  fixing one half of #17266.
* Updated documentation fixes other half of #17266.
* password_hash does not hard-code the salt-length, which fixes bcrypt
  in connection with passlib.
  bcrypt requires a salt with length 22, which fixes #25347
* Salts are only generated by ansible when using crypt.crypt.
  Otherwise passlib generates them.
* Avoids deprecated functionality of passlib with newer library versions.
* When no rounds are specified for sha256/sha256_crypt and sha512/sha512_crypt
  always uses the default values used by crypt, i.e. 5000 rounds.
  Before when installed passlibs' defaults were used.
  passlib changes its defaults with newer library versions, leading to non
  idempotent behavior.

  NOTE: This will lead to the recalculation of existing hashes generated
        with passlib and without a rounds parameter.
        Yet henceforth the hashes will remain the same.
        No matter the installed passlib version.
        Making these hashes idempotent.

Fixes #15326
Fixes #17266
Fixes #25347 except bcrypt still uses 2a, instead of the suggested 2b.

mfuchs commented:
Finally had time to work on this again.

Further changes:
* Always use the crypt rounds defaults when no rounds were specified manually. This will lead to new hashes for vars_prompt. They will be idempotent from then on.
* Fixes #25347 except bcrypt still uses 2a, instead of the suggested 2b.

To discuss:
* what about vars_prompt? Should there be a way to specify the rounds?
* @dharmabumstead I can't use the nicer tests as not every machine has the needed libraries installed. That is why I have these ugly checks for passlib and darwin. I'd be very grateful if you have a better idea, because this is really ugly. :(

dharmabumstead commented:
Closing due to inactivity. Please feel free to reopen and rebase.

Akasurde commented:
@The-Wabbit @pilou- @bcoca Could you please review this ?

BarbzYHOOL commented:
Every commands takes 15sec until this is merged, this is crazy. Users passwords are set with that horrible setting "rounds=656000"

ref: https://serverfault.com/questions/909248/can-you-set-the-number-of-rounds-that-ansibles-password-hash-jinja2-filter-uses

pilou- commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -253,42 +254,19 @@ def get_hash(data, hashtype='sha1'):
-
-    return None
+    hashtype = passlib_mapping.get(hashtype, hashtype)
+    return passlib_or_crypt(password, hashtype, error_type=AnsibleFilterError, **settings)
```
Should not `None` be returned when `hashtype` is unknown instead of raising `AnsibleFilterError` ? I mean without this change, `None` is returned when `hashype` is unknown, is there a reason to change that ?

mfuchs replied:
Because I think otherwise there could be unexpected behaviour. A typo in the hash-type or a unsupported type by crypt should not lead to the function being successful. Instead the issue should be pointed out right away.
This also makes the function more consistent with the old do_encrypt.py of encrypt.py.

abadger replied:
Typically, our code uses try: except to catch and transform exception rather than making explicit passing of exception types part of a function's signature.  There are drawbacks to try: except (for instance, changing the traceback on python2) but I think we should stick with that idiom.

mfuchs replied:
I modified it to always raise AnsibleError.
So also in the filter context (password_hash).
Please tell me whether I should raise AnsibleFilterError there (via the steps you suggested -> https://nedbatchelder.com/blog/200711/rethrowing_exceptions_in_python.html)?

pilou- commented on lib/ansible/utils/encrypt.py:
```diff
@@ -28,36 +33,161 @@
-                salt = to_bytes(salt, encoding='ascii', errors='strict')
+    def _clean_salt(self, settings):
+        if "salt" in settings:
+            if self.crypt_algo._salt_is_bytes:
```
Should not `isinstance(self.crypt_algo, HasRawSalt)` be used instead of accessing a [private helper](https://bitbucket.org/ecollins/passlib/src/849ab1e6b5d4ace4c727a63d4adec928d6d72c13/passlib/utils/handlers.py?at=default&fileviewer=file-view-default#handlers.py-1236) ?

mfuchs replied:
That is a good idea.
I just tried and there seem to be issues:
```python
    assert type(passlib.hash.bcrypt) is abc.ABCMeta
```
It appears to me that some python magic is done in passlib/ifc.py with PasswordHash which seems to kinda break your approach. :(

pilou- replied:
Indeed, `issubclass` should be used:
```
In [1]: issubclass(getattr(passlib.hash, 'sha256_crypt'), HasRawSalt)
Out[1]: False

In [2]: issubclass(getattr(passlib.hash, 'pbkdf2_sha256'), HasRawSalt)
Out[2]: True
```

pilou- commented on lib/ansible/utils/encrypt.py:
```diff
@@ -28,36 +33,161 @@
+            result = self.crypt_algo.encrypt(secret, **settings)
         else:
-            result = crypt.encrypt(result)
+            raise self.error_type("installed passlib version not supported")
```
Error message should contain `passlib.__version__`

pilou- commented on lib/ansible/utils/encrypt.py:
```diff
@@ -28,36 +33,161 @@
+        return to_text(result, errors='strict')
+
+
+def crypt_hash(secret, algorithm, error_type, **settings):
```
`crypt_hash` and `passlib_hash` methods don't seem useful: `passlib_or_crypt` could use `CryptHash`/`PasslibHash` directly.

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,123 @@
+        encrypt.PASSLIB_AVAILABLE = self.orig
+
+
+class TestEncrypt(unittest.TestCase):
```
New unit tests must be `pytest` based (not `unittest` based).

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,123 @@
+
+class TestEncrypt(unittest.TestCase):
+    def assertHash(self, expected, secret, algorithm, **settings):
+        if IS_DARWIN:
```
With shippable, `IS_DARWIN` will always be `False` (unit tests aren't executed using OS X).

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,123 @@
+        else:
+            self.assertEquals(encrypt.crypt_hash(secret, algorithm, AnsibleFilterError, **settings), expected)
+
+        if encrypt.PASSLIB_AVAILABLE:
```
Should not `passlib` be available ?

mfuchs replied:
At least last time the tests were run passlib was not available on all test-platforms.

mfuchs commented:
@pilou- thanks for your comments! I'll check through them and then will push a cleaned-up and rebased version the next days. :)

BarbzYHOOL commented:
Awesome, I hope this gets merged

mfuchs commented:
@pilou- can you please re-review my changes? :)

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,117 @@
+            encrypt.PasslibHash(algorithm, AnsibleFilterError).hash(secret, **settings)
+
+
+def test_encrypt_with_rounds():
```
Should not an algorithm using a _raw salf_ be tested ?

pilou- commented on lib/ansible/utils/encrypt.py:
```diff
@@ -33,31 +38,137 @@
+        if salt is None:
+            if salt_size is None:
+                salt_size = self.algo_data.salt_size
+            r = random.SystemRandom()
```
Should not `ansible.utils.encrypt.random_password` be reused here ([example](https://github.com/ansible/ansible/blob/adc0efe10c5be833b5138e1d8b0d0316f87c5241/lib/ansible/plugins/lookup/password.py#L207)) ?

mfuchs replied:
I did some modifications:
* Added encrypt.random_salt which is called by password._random_salt and also used in encrypt.py
* random_salt does not do a to_text. I am not sure if that is actually needed.
  Before to_text was indirectly used via _gen_candiate_chars

mfuchs committed:
Applies review comments.

pilou- commented:
Network failures on Fedora/RHEL/CentOS:
* RHE/7.4/1
```
2018-08-09 15:52:25 TASK [yum : download an rpm] ***************************************************
2018-08-09 15:52:25 fatal: [testhost]: FAILED! => {"changed": false, "dest": "/tmp/fpaste.rpm", "msg": "Request failed", "response": "HTTP Error 404: Not Found", "state": "absent", "status_code": 404, "url": "http://download.fedoraproject.org/pub/epel/7/x86_64/Packages/f/fpaste-0.3.7.4.1-1.el7.noarch.rpm"}
```
* CentOS 6
```
2018-08-09 15:52:56 TASK [yum : download an rpm] ***************************************************
2018-08-09 15:52:56 fatal: [testhost]: FAILED! => {"changed": false, "dest": "/tmp/fpaste.rpm", "msg": "Request failed", "response": "HTTP Error 404: Not Found", "state": "absent", "status_code": 404, "url": "http://download.fedoraproject.org/pub/epel/7/x86_64/Packages/f/fpaste-0.3.7.4.1-1.el7.noarch.rpm"}
```

mattclay commented:
I've fixed the yum tests and restarted CI for this PR.

pilou- commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -204,12 +204,7 @@ def _gen_candidate_chars(characters):
 
 
 def _random_salt():
```
Could not `_random_salt` method be deleted ?

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,122 @@
+
+def test_password_hash_filter():
+    if not encrypt.PASSLIB_AVAILABLE:
+        return
```
`pytest.skip("passlib not available")` should be used ?

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,122 @@
+
+    assert encrypt.do_encrypt("123", "crypt16", salt="12") == "12pELHK2ME3McUFlHxel6uMM"
+
+    with passlib_off():
```
Should not this block moved at the beginning of the test method (in order to be executed even when passlib isn't available) ?

mfuchs committed:
Implements review comments.

* random_salt is solely handled by encrypt.py.
  There is no _random_salt function there anymore.
  Also the test moved to test_encrypt.py.
* Uses pytest.skip when passlib is not available, instead of a silent return.
* More checks are executed when passlib is not available.

mfuchs commented:
@pilou- thanks for your input! I pushed another commit to make it easier for you to check the diff.

Should I squash them all toghether and rebase them on most recent devel?

pilou- commented on test/units/utils/test_encrypt.py:
```diff
@@ -0,0 +1,130 @@
+        with pytest.raises(AnsibleError):
+            encrypt.do_encrypt("123", "crypt16", salt="12")
+
+    if not encrypt.PASSLIB_AVAILABLE:
```
It seems best to split the `test_do_encrypt`: one part without passlib (first part of the current method), one with (this one will use `pytest.skip`). Sorry, my previous comments were not clear :-/

mfuchs replied:
Indeed. I realised that myself just when I left for lunch. :)

mfuchs committed:
Moves tests that require passlib into their own test-function.

pilou- reviewed (approved):
shipit

abadger commented on lib/ansible/plugins/filter/core.py:
```diff
@@ -254,42 +249,19 @@ def get_hash(data, hashtype='sha1'):
 
 
-def get_encrypted_password(password, hashtype='sha512', salt=None):
+def get_encrypted_password(password, hashtype='sha512', salt=None, **settings):
```
I'm not a fan of passing opaque data structures (```settings```) around.  I'd much rather that these parameters were explicitly listed. (In the other functions as well).

abadger commented on lib/ansible/utils/encrypt.py:
```diff
@@ -33,31 +40,140 @@
+            return None
+        return settings.get('rounds')
+
+    def _encrypt(self, secret, salt, rounds):
```
Don't use the name _encrypt for hashing... I know we already have a publically available function named do_encrypt which really does hashing but there's no sense making that problem worse.

abadger commented on lib/ansible/utils/encrypt.py:
```diff
@@ -33,31 +40,140 @@
+            raise self.error_type("crypt.crypt does not support '%s' algorithm" % self.algorithm)
+        self.algo_data = self.algorithms[algorithm]
+
+    def hash(self, secret, **settings):
```
Note: organizationally, our code is largely, callers near the bottom of the file, callees near the top.  (Like C code).

abadger commented on lib/ansible/utils/encrypt.py:
```diff
@@ -33,31 +40,140 @@
+        if algo_data and algo_data.implicit_rounds and "rounds" not in settings:
+            # Ensure passlib behaves the same as crypt in case no rounds were specified.
+            # That means instead of using the passlib default for the rounds the crypt default is used.
+            settings["rounds"] = algo_data.implicit_rounds
```
For _clean_salt and _clean_rounds, just pass in the one setting you care about and then return the modified version of it.  Change the container in the caller.  functional programming like that is easier to test and easier to reason about.

abadger commented:
Okay, I took a look.  the logic here looks fine but there's some coding style/API that I think should be changed.  I'll be busy with the new release so if you get some time to work on this but I don't notice, feel free to ping me on IRC (abadger1999 on irc.freenode.net, the #ansible-devel channel).

mfuchs committed:
Instead of passing in the error type passlib_or_crypt always uses AnsibleError.

The password_hash filter then transforms the AnsibleError to an AnsibleFilterError.

mfuchs committed:
Get rid of the settings dict, instead be specific about the function parameters.

mfuchs committed:
Reorders the code, so that the callee is on top.

mfuchs commented:
@abadger all comments should be handled now. Thanks for your input!

Just in case the rethrowing with keeping the traceback works differently in Python 2 and Python3.
Thus the code is quite ugly.

If this is a common use case then it should be moved to a better place (a function), something like reraise_as.
Of course reraise_as should then modify the traceback to not be part of it.

abadger commented:
For the raise syntax issues... we do bundle six as ansible.module_utils.six... six has a couple raise functions but I'm not sure if they're precisely what you need or not.  (Also, if one of them can do what ned Batchelder's recipe does, sorry for not thinking of it earlier... I'll blame lack of sleep for not remembering those if so ;-)

I'll look in on this in the morning.

mfuchs committed:
Uses the six library to reraise the exception.

mfuchs commented:
Some tests fail.
I checked and at least test/integration/targets/vars_prompt/vars_prompt-5.yml fails.
The reason is that it expects `re.compile('"password": "\\$6\\$rounds=')` in the password-output.
Yet since no rounds are provided the default of crypt is used leading to a output like

> "password": "$6$jESIyad4F08hP3Ta$hrnwEbDk9uSg1IURdckNLBy9xoeO1y1peRDBHsdmwzlWv1u9Cs0qasiy7THQ88dnzC0f4PaFX4ySUCcyT5emn0"

This change was implemented to be more consistent overall, thus having consistency between vars_prompt and password_hash (no matter if passlib is installed or not) and not depending on the always changing defaults of passlib.

abadger commented:
Makes sense to me.  Go ahead and change the test.

mfuchs committed:
Fixes integration test.

When no rounds are provided the defaults of crypt are used.
In that case the rounds are not part of the resulting MCF output.

mfuchs commented:
Since the test is rather new I rebased on current devel + adapted the integration tests.

abadger committed:
porting guide and changelog for hashing changes

abadger commented:
Merged to devel for the 2.7.0 release.  Thanks @mfuchs

BarbzYHOOL commented:
awesome
